File: server/conf.js
// ponytail: conf + DB paths come from CLI flags (-c / -D). Fallback: next to the running
// binary (process.execPath when packaged, dirname(import.meta.url) when run via `node`).
import fs from 'node:fs'
import path from 'node:path'
import os from 'node:os'
import crypto from 'node:crypto'

const md5 = (s) => crypto.createHash('md5').update(s).digest('hex')

// ponytail: unified data dir. One subdir across all three launchers (cli / desktop / standalone server).
// Always `~/.cj-nantianmen/` on every OS. User explicitly chose this over platform-specific appData.
function defaultBaseDir() {
  return path.join(os.homedir(), '.cj-nantianmen')
}

// ponytail: lazy init. setPaths() must be called by server/index.js BEFORE loadConf().
let _confPath = null
let _dbPath = null

export function setPaths({ confPath, dbPath } = {}) {
  if (confPath) _confPath = path.resolve(confPath)
  if (dbPath) _dbPath = path.resolve(dbPath)
}

// ponytail: paths resolved at first read. setPaths() influences defaults for the very first call.
let _resolvedBase = null
function base() {
  if (!_resolvedBase) _resolvedBase = defaultBaseDir()
  return _resolvedBase
}

let _resolvedConfPath = null
function confPathLazy() {
  if (_resolvedConfPath) return _resolvedConfPath
  _resolvedConfPath = _confPath || path.join(base(), 'nantianmen-conf.json')
  return _resolvedConfPath
}

let _dirsReady = false
function ensureDirs() {
  if (_dirsReady) return
  _dirsReady = true
  // ponytail: home dir may not exist yet on first run (e.g. fresh VM).
  try { fs.mkdirSync(path.dirname(getConfPath()), { recursive: true }) } catch {}
  try { fs.mkdirSync(path.dirname(getDefaultDbPath()), { recursive: true }) } catch {}
}

export function getConfPath() { ensureDirs(); return _confPath || confPathLazy() }
export function getConfDir() { return path.dirname(getConfPath()) }
export function getDefaultDbPath() { ensureDirs(); return _dbPath || path.join(base(), 'nantianmen.db') }

function defaultDbPath() { return getDefaultDbPath() }

export const DEFAULT_CONF = () => ({
  initialized: true,
  server_host: '127.0.0.1',
  server_port: 38271,
  password: '',
  salt: '',
  log_enabled: false,
  // ponytail: outbound proxy for upstream LLM calls. `system` uses env HTTPS_PROXY/ALL_PROXY,
  // `direct` bypasses any system proxy, `custom` uses `proxy_url` (e.g. http://host:port or socks5://host:port).
  proxy: 'system',
  proxy_url: '',
  database: { type: 'sqlite3', path: defaultDbPath() },
})

let _conf = null

export function loadConf() {
  const cp = getConfPath()
  if (fs.existsSync(cp)) {
    _conf = JSON.parse(fs.readFileSync(cp, 'utf8'))
  } else {
    _conf = { ...DEFAULT_CONF() }
    if (!_conf.salt) _conf.salt = randomSalt()
    // ponytail: bootstrap password so admin auth has something to verify against.
    // Default 'admin' md5 + salt → stored hash. Users can change later via /password/change.
    if (!_conf.password) {
      _conf.password = md5(md5('admin') + _conf.salt)
    }
    saveConf()
  }
  // ponytail: relative DB paths always resolve against conf dir (exeDir in portable).
  // First-run path: conf didn't exist → DEFAULT_CONF had raw "./nantianmen.db" or setPaths()'s path → must resolve.
  if (_conf.database && _conf.database.path && !path.isAbsolute(_conf.database.path)) {
    _conf.database.path = path.resolve(getConfDir(), _conf.database.path)
  }
  return _conf
}

export function getConf() {
  if (_conf === null) loadConf()
  return _conf
}

export function updateConf(patch) {
  Object.assign(_conf, patch)
  saveConf()
  return _conf
}

export function saveConf() {
  fs.writeFileSync(getConfPath(), JSON.stringify(_conf, null, 2))
}

export function randomSalt() {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'
  let s = ''
  for (let i = 0; i < 6; i++) s += chars[Math.floor(Math.random() * chars.length)]
  return s
}

File: server-go/internal/conf/conf.go
package conf

import (
	"crypto/md5"
	"crypto/rand"
	"encoding/json"
	"fmt"
	"math/big"
	"os"
	"path/filepath"
)

var (
	ConfPath string
	DbPath   string
	conf     *Config
)

type Config struct {
	Initialized         bool        `json:"initialized"`
	ServerHost          string      `json:"server_host"`
	ServerPort          int         `json:"server_port"`
	Password            string      `json:"password"`
	Salt                string      `json:"salt"`
	LogEnabled          bool        `json:"log_enabled"`
	LogRotationEnabled  bool        `json:"log_rotation_enabled"`
	LogRotationMax      int         `json:"log_rotation_max"`
	Proxy               string      `json:"proxy"`
	ProxyURL            string      `json:"proxy_url"`
	Database            DatabaseCfg `json:"database"`
	UIFilters           interface{} `json:"ui_filters"`
	WindowState         interface{} `json:"window_state"`
}

type DatabaseCfg struct {
	Type string `json:"type"`
	Path string `json:"path"`
}

func DefaultBaseDir() string {
	h, _ := os.UserHomeDir()
	return filepath.Join(h, ".cj-nantianmen")
}

func GetConfDir() string {
	return filepath.Dir(GetConfPath())
}

func GetConfPath() string {
	if ConfPath != "" {
		return ConfPath
	}
	ConfPath = filepath.Join(DefaultBaseDir(), "nantianmen-conf.json")
	return ConfPath
}

func GetDefaultDbPath() string {
	if DbPath != "" {
		return DbPath
	}
	DbPath = filepath.Join(DefaultBaseDir(), "nantianmen.db")
	return DbPath
}

func DefaultConf() Config {
	return Config{
		Initialized: true,
		ServerHost:  "127.0.0.1",
		ServerPort:  38271,
		Password:    "",
		Salt:        "",
		LogEnabled:  false,
		Proxy:       "system",
		ProxyURL:    "",
		Database: DatabaseCfg{
			Type: "sqlite3",
			Path: GetDefaultDbPath(),
		},
	}
}

func RandomSalt() string {
	const chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
	s := make([]byte, 6)
	for i := range s {
		n, _ := rand.Int(rand.Reader, big.NewInt(int64(len(chars))))
		s[i] = chars[n.Int64()]
	}
	return string(s)
}

func Md5(s string) string {
	return fmt.Sprintf("%x", md5.Sum([]byte(s)))
}

func LoadConf() Config {
	cp := GetConfPath()
	data, err := os.ReadFile(cp)
	if err == nil {
		json.Unmarshal(data, &conf)
		if conf != nil {
			// Resolve relative DB path
			if conf.Database.Path != "" && !filepath.IsAbs(conf.Database.Path) {
				conf.Database.Path = filepath.Join(GetConfDir(), conf.Database.Path)
			}
			return *conf
		}
	}
	c := DefaultConf()
	if c.Salt == "" {
		c.Salt = RandomSalt()
	}
	if c.Password == "" {
		c.Password = Md5(Md5("admin") + c.Salt)
	}
	conf = &c
	SaveConf()
	return c
}

func GetConf() Config {
	if conf == nil {
		return LoadConf()
	}
	return *conf
}

func UpdateConf(patch map[string]interface{}) {
	if conf == nil {
		c := LoadConf()
		conf = &c
	}
	// Simple merge
	data, _ := json.Marshal(patch)
	json.Unmarshal(data, conf)
	SaveConf()
}

func SaveConf() {
	if conf == nil {
		return
	}
	data, _ := json.MarshalIndent(conf, "", "  ")
	os.MkdirAll(filepath.Dir(GetConfPath()), 0700)
	os.WriteFile(GetConfPath(), data, 0600)
}

func SetPaths(cfgPath, dbPath string) {
	if cfgPath != "" {
		ConfPath = cfgPath
	}
	if dbPath != "" {
		DbPath = dbPath
	}
}
